katdal/caldata.py

```diff
@@ -0,0 +1,383 @@
+                                                                              kind='linear',
+                                                                              fill_value='extrapolate',
+                                                                              assume_sorted=True,
+                                                                              )
+
+    # to interpolate you need >1 timestamps, else return only the single value
+    # if values.shape[0] > 1:
+    if timestamps.size > 1:
+        kind = kwargs.get('kind', 'linear')  # default if none given
+        if np.iscomplexobj(values) and kind not in ['zero', 'nearest']:
+            interp = ComplexInterpolate1D
+        else:
+            interp = scipy.interpolate.interp1d
+        return interp(timestamps,
+                      values,
+                      axis=0,
+                      fill_value='extrapolate',
+                      assume_sorted=True,
+                      **kwargs)
+    return values
+
+
+def _cal_setup(katdal_obj):
+    katdal_obj._cal_pol_ordering = _get_cal_pol_ordering(katdal_obj)
+    katdal_obj._cal_ant_ordering = _get_cal_antlist(katdal_obj)
+    katdal_obj._data_channel_freqs = katdal_obj.channel_freqs
+    katdal_obj._delay_to_phase = (-2j * np.pi * katdal_obj._data_channel_freqs)[np.newaxis, :, np.newaxis, np.newaxis]
+    # baselines from the data files are given as <ant><pol>
+    # for indexing this has to be swapped to <pol><ant>
+    katdal_obj._cp_lookup = [[(katdal_obj._cal_pol_ordering[prod[0][-1]], katdal_obj._cal_ant_ordering.index(prod[0][:-1]),),
+                              (katdal_obj._cal_pol_ordering[prod[1][-1]], katdal_obj._cal_ant_ordering.index(prod[1][:-1]),)]
+                             for prod in katdal_obj.corr_products]
+    katdal_obj._cp_lookup = np.asarray(katdal_obj._cp_lookup, dtype=int)
+    # interpolation function over available solutions
+    for key in katdal_obj._cal_solns.keys():
+        try:
+            katdal_obj._cal_solns[key]['interp'] = _get_cal_product('cal_product_' + key,
+                                                                    katdal_obj,
+                                                                    kind=katdal_obj._cal_solns[key]['kind'],
+                                                                    )
+        except CalibrationReadError:
+            raise
+        except:  # no delay cal solution from telstate
+            # should raise a warning
+            raise
+    return katdal_obj
+
+
+def applycal(katdal_obj):
+    """
+    Apply the K, B, G solutions to visibilities at provided timestamps.
+    Optionally recompute the weights as well.
+
+    Returns
+    =======
+    katdal_obj: containing vis and weights (optional) with cal solns applied.
+    """
+    initcal = lambda kind: {'interp': None, 'solns': None, 'kind': kind}
```

Comment:
replaced with a def
had a look at the attrs modules, looks very useful, but will add functionality to the new repository to prevent pollution to katdal for a single need